srv/modules/runners/upgrade.py

```diff
@@ -109,6 +112,131 @@ def check():
     return ret
 
 
+def _sort_nodes_by_role(nodes, roles):
+    """
+    Given a list of nodes, return the list sorted first by role in upgrade order
+    (master, then mon/mgr, then mds, then storage, ...), and secondly by name.
+
+    The roles parameter is a dict mapping _all_ node names to a list of their
+    assigned roles (i.e. `salt '*' pillar.get roles`), so that this function can
+    actually figure out what roles are assigned to each node in the nodes list.
+    """
+    nodes_sorted = []
+    for role in ['master', 'mon', 'mgr', 'mds', 'storage', 'rgw', 'igw', 'ganesha']:
+        role_nodes = [node for node in roles if role in roles[node] and node in nodes]
+        role_nodes.sort()
+        nodes_sorted.extend([node for node in role_nodes if node not in nodes_sorted])
+    return nodes_sorted
+
+
+def _print_nodes_to_upgrade(releases, newest, roles):
+    """
+    Prints a list of nodes currently on the newest release, followed by an ordered
+    list of which nodes need upgrading next.
+
+    releases is a dict mapping node names to software versions
+    newest is the newest installed version
+    roles is a dict as used by _sort_nodes_by_role()
+    """
+    upgraded_nodes = _sort_nodes_by_role([node for node in releases if releases[node] == newest], roles)
+    nodes_to_upgrade = _sort_nodes_by_role([node for node in releases if releases[node] < newest], roles)
+
+    print("Nodes running these software versions:")
+    for node in upgraded_nodes:
+        print("  {} (roles: {})".format(node, ", ".join(roles[node])))
+    print("")
+
+    print("Nodes running older software versions must be upgraded in the following order:")
+    i = 1
+    for node in nodes_to_upgrade:
+        print("{:4}: {} (roles: {})".format(i, node, ", ".join(roles[node])))
+        i += 1
+    print("")
+
+
+def status():
+    """
+    Check status of upgrade.  This is similar to status.report, but if the
+    base OS version or ceph version isn't identical on all nodes, it will
+    advise which node(s) should be upgraded next.
+    """
+    __opts__ = salt.config.client_config('/etc/salt/master')
+    __grains__ = salt.loader.grains(__opts__)
+    __opts__['grains'] = __grains__
+    __utils__ = salt.loader.utils(__opts__)
+    __salt__ = salt.loader.minion_mods(__opts__, utils=__utils__)
+
+    os_codename, _, ceph_version = __utils__['status.get_sys_versions']()
```

Comment:
Isn't that what you're aiming to get from the `status` runner?

Reply:
Not sure I understand the question -- I moved the guts of the status runner to a separate utils module so both status.report and upgrade.status could rely on the same core code.